### PR TITLE
Added support for ABGR in texc, in order to skip a java data copy loop

### PR DESCRIPTION
**Pull request by JCash**
Saves ~20-25% in the case of the spineboy atlas. (~100-125ms for the copy step)

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/TexcLibrary.java [hunk 1]
```diff
@@ -60,23 +60,24 @@ static void addToPath(String variable, String path) {
...
-
-        public static int RGBA_ETC2         = 11;
-        public static int RGBA_ASTC_4x4     = 12;
-
-        public static int RGB_BC1           = 13;
-        public static int RGBA_BC3          = 14;
-        public static int R_BC4             = 15;
-        public static int RG_BC5            = 16;
-        public static int RGBA_BC7          = 17;
+        public static int B8G8R8A8          = 3;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/TexcLibrary.java]: Added.
The rest is just the increased enum values

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java [hunk 2]
```diff
@@ -177,8 +187,53 @@ else if (componentCount == 3)
...
 
+    private static ByteBuffer getByteBuffer(BufferedImage bi)
+    {
+        ByteBuffer byteBuffer;
+        DataBuffer dataBuffer = bi.getRaster().getDataBuffer();
+
+        if (dataBuffer instanceof DataBufferByte) { // This is the usual case, where data is simply wrapped
+            byte[] pixelData = ((DataBufferByte) dataBuffer).getData();
+            byteBuffer = ByteBuffer.wrap(pixelData);
+        }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:198]: Huge gain, we simply wrap the image data, as it's already in byte layout.
(Although in ABGR)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java [hunk 3]
```diff
@@ -187,22 +228,10 @@ else if (componentCount == 3)
...
 
-            int count = Math.min(height - y, 512);
 
-            int[] rasterData = new int[count*width*4];
-            image.getRaster().getPixels(0, y, width, count, rasterData);
-            for (int i = 0; i < rasterData.length; ++i) {
-                buffer.put((byte) (rasterData[i] & 0xff));
-            }
-        }
+        ByteBuffer buffer_input = getByteBuffer(image);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:233]: The previous code was from 2015, and I'd like to think that the Linux machines are now all around beefier.

Also, even _before_ that fix, it was a big copy.
Now, it's just a "wrap" of the same memory, before we pass it into the texc.cpp.

We'll try this first, and if it becomes a problem, we'll look into it again.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java (hunk 2 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:198]: isn't it better to prepare texture in bob and that use just memcpy in runtime? I mean we always try to make everything we can during the build process

**Review by britzl — APPROVED** [review]

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java (hunk 2 above)
Comment (JCash) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:198]: Not sure what you  mean?
This _is_ the build process?
We pass the data to the c++ code (texc.cpp) for processing.
We don't do the compression etc in Java.

**Pull request merged**